src/typechecker.erl

```diff
@@ -1412,8 +1412,19 @@ type_check_expr(Env, {record, Anno, Record, Fields}) ->
     RecTy    = type(record, [{atom, erl_anno:new(0), Record}]),
     {VB, Cs} = type_check_fields(Env, Record, Anno, Fields),
     {RecTy, VB, Cs};
-type_check_expr(_Env, {record_index, _, _Record, _Field}) ->
-    {type(integer), #{}, constraints:empty()};
+type_check_expr(Env, {record_index, Anno, Name, FieldNameAtom}) ->
+    %% we still fetch record field index even if infer=false
+    %% to be able to report undefined record or record field.
+    %% ("we do our best for the customer")
```

Comment:
In other words, we're checking things that we don't need, to be able to report an error that the compiler would also report.

In principle, I think it's the wrong approach, but in this case, I don't see that there might be any performance impact and the complexity of the code is the same either way, so keep it!

Reporting an exact type for record index when inference is on is very good!